src/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Compression
+{
+    public partial class BrotliStream : Stream
+    {
+        private BrotliDecoder _decoder;
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateParameters(buffer, offset, count);
+            return ReadCore(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> destination)
+        {
+            if (GetType() != typeof(BrotliStream))
+            {
+                // BrotliStream is not sealed, and a derived type may override Read(byte[], int, int) without also
+                // overriding Read(Span<byte>). In that case, this Read(Span<byte>) overload
+                // should use the behavior of Read(byte[],int,int) overload.
+                return base.Read(destination);
+            }
+            else
+            {
+                return ReadCore(destination);
+            }
+        }
+
+        internal int ReadCore(Span<byte> destination)
+        {
+            if (_mode != CompressionMode.Decompress)
+                throw new InvalidOperationException(SR.BrotliStream_Compress_UnsupportedOperation);
+            EnsureNotDisposed();
+            int totalWritten = 0;
+            Span<byte> source;
+            OperationStatus lastResult = OperationStatus.DestinationTooSmall;
+            while (destination.Length > 0 && (lastResult == OperationStatus.DestinationTooSmall || lastResult == OperationStatus.NeedMoreData))
+            {
+                int bytesConsumed = 0;
+                int bytesWritten = 0;
+
+                if (lastResult == OperationStatus.NeedMoreData)
+                {
+                    int readBytes = 0;
+                    int iter = 0;
+                    while (readBytes < _buffer.Length && ((iter = _stream.Read(_buffer, readBytes, _buffer.Length - readBytes)) > 0))
+                    {
+                        readBytes += iter;
+                        if (readBytes > _buffer.Length)
+                        {
+                            // The stream is either malicious or poorly implemented and returned a number of
+                            // bytes larger than the buffer supplied to it.
+                            throw new InvalidDataException(SR.BrotliStream_Decompress_InvalidStream);
+                        }
+                    }
+                    if (readBytes <= 0)
+                    {
+                        break;
+                    }
+                    source = new Span<byte>(_buffer, 0, readBytes);
+                }
+
+                lastResult = _decoder.Decompress(source, destination, out bytesConsumed, out bytesWritten);
```

Comment:
I'm a little surprised this compiles; I'd have thought this would result in an error about `source` potentially not being initialized.  I guess not?

Reply:
My understanding is that since Span is a struct, the declaration on line 42 stackallocs a Span using the "default" variables which cause the Span to effectively be treated as being Span<byte>.Empty. 

It causes no errors when the above loop performs no iterations and behaves as if I had declared it on line 42 as ```Span<byte> source = Span<byte>.Empty```. It may actually be less performant though since I'm constructing a copy that isn't necessary, so I'll change it to use ```Empty```.